Wrappers/Python/cil/optimisation/functions/SAGFunction.py

```diff
@@ -0,0 +1,180 @@
+
+
+    
+
+    """
+
+    def __init__(self, functions,  sampler=None):
+        self._list_stored_gradients = None
+        self._full_gradient_at_iterate = None
+        self._warm_start_just_done=False
+        self._sampled_grad=None
+        
+        super(SAGFunction, self).__init__(functions, sampler)
+
+
+        
+
+    def approximate_gradient(self, x, function_num,  out=None):
+        """ SAG approximate gradient, calculated at the point :math:`x` and updated using the function index given by `function_num`.  
+
+        Parameters
+        ----------
+        x: element in the domain of the `functions`
+
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+
+        """
+        
+        
+        if self._list_stored_gradients is None: # Initialise the stored gradients on the first call of gradient unless using warm start.  
+            self._list_stored_gradients = [
+                0*x for fi in self.functions]
+            self._full_gradient_at_iterate = 0*x
+            
+        
+        if self.function_num >= self.num_functions or self.function_num<0 : #check the sampler and raise an error if needed
+            raise IndexError(
+                'The sampler has outputted an index larger than the number of functions to sample from. Please ensure your sampler samples from {0,1,...,len(functions)-1} only.')
+
+            
+        #Calculate the gradient of the sampled function at the current iterate 
+        self._sampled_grad = self.functions[function_num].gradient(x)
+
+        
+        #Calculate the difference between the new gradient of the sampled function and the stored one
+        self._stochastic_grad_difference = self._sampled_grad.sapyb(
+            1., self._list_stored_gradients[function_num], -1.)
+
+        #Calculate the  approximate gradient
+        out =self._update_approx_gradient(out)
+
+        #Update the stored gradients 
+        self._list_stored_gradients[function_num].fill(
+            self._sampled_grad)
+        
+        #Calculate the stored full gradient
+        self._full_gradient_at_iterate.sapyb(
+            1., self._stochastic_grad_difference, 1., out=self._full_gradient_at_iterate)
+
+        return out
+    
+    def _update_approx_gradient(self, out):
+        """Internal function used to differentiate between the SAG and SAGA calculations. This is the SAG approximation: """
+        if out is None:
+            out = self._stochastic_grad_difference.sapyb(
+                1., self._full_gradient_at_iterate, 1.)
+        else:
+            self._stochastic_grad_difference.sapyb(
+                1., self._full_gradient_at_iterate, 1., out=out)
+
+        return out 
+    
+    def warm_start_approximate_gradients(self, initial):
```

Comment:
If we want to run sth like
```
prox_saga = ISTA(initial = FBP_recon, f = f_saga, g = G ....)
```
or
```
prox_saga = ISTA(initial = sgd_2epochs, f = f_saga, g = G ....)
```
and forget to 
```
f_saga.warm_start_approximate_gradients(FBP_recon)
```
then we will have a problem. Also, by changing one input for different stochastic estimators, e.g, svrg, lsvrg, we need to be careful not to forget
```
f_svrg.warm_start_approximate_gradients(FBP_recon)
f_lsvrg.warm_start_approximate_gradients(FBP_recon)
```
objective values will not be affected but we are going to have issues with the iterates.

Reply:
Hi @epapoutsellis  - currently I don't have a `warm_start_approximate_gradients` for SVRG and LSVRG because the first iteration always takes a snapshot anyway. 

The question here is if we warn start by default or if we the user call this function as an additional step. I know we have discussed the pros and cons before and have changed our minds before.  The idea of making the user do this additional step is to make it clearer what is going on under the hood, but yes, there is the risk they don't go on to initialise with the same value they called warm start on.

Reply:
Any decision on this? Do they need to add a new line so the initial of the algorithm is the same initial for the gradient point of the stochastic estimator function?

Reply:
Had a chat with @paskino and discussed how the user would interact with this. As it stands at the moment the user would have to:
'''python
f_saga=SAGFunction(functions)
f_saga.warm_start_approximate_gradients(FBP_recon)
prox_saga = ISTA(initial = FBP_recon, f = f_saga, g = G ....)  
'''

They should make sure that the initial they give to the algorithm is the same they give to warm start the approximate gradient function. 

I think we are happy with this as it makes it clear to the user what is happening.

Reply:
Ok fine with me. But please note with this option we allow the user to set two different points a) to compute the gradient at x1 (descent direction) b) to compute the objective at another point x2. This does not make any mathematical sense.

Reply:
Thanks Vaggelis! I will make sure this is written clearly in the documentation